### PR TITLE
Add simulated test for non-zero spinel iid.

### PR DESCRIPTION
**Pull request by parag-silabs**
- Added support in simulated PAL (radio.c) to tag correct spinel iid for received packets.
- Modified script/test to build a target for non-zero spinel iid tests. 
- Update CMakeLists.txt for openthread-simulation and openthread-platform-utils libraries to build with - OPENTHREAD_RADIO symbol when OT_MULTIPAN_RCP is enabled. 
- Added 'expect' test (posix-rcp-non-zero-iid.exp) for non-zero spinel iid.

Expect test will run as a part of git posix workflow under 'expect-linux' job.

### CODE REVIEW

File: examples/platforms/simulation/radio.c [hunk 1]
```diff
@@ -299,25 +308,49 @@ static void ReverseExtAddress(otExtAddress *aReversed, const otExtAddress *aOrig
...
 static bool hasFramePending(const otRadioFrame *aFrame)
 {
     bool         rval = false;
     otMacAddress src;
+    uint8_t      iid;
 
     otEXPECT_ACTION(sSrcMatchEnabled, rval = true);
     otEXPECT(otMacFrameGetSrcAddr(aFrame, &src) == OT_ERROR_NONE);
+    iid = getIidFromFrame(aFrame);
+    assert(iid != INVALID_VALUE);
```
Comment (hemanth-silabs) [inline examples/platforms/simulation/radio.c:341]: No need to assert here. There might be tests (now or in future), where we validate that packets with wrong dest panid are handled. We can either drop it or pass it with bcast iid.

File: examples/platforms/simulation/radio.c [hunk 2]
```diff
@@ -1083,6 +1119,10 @@ void radioProcessFrame(otInstance *aInstance)
     otEXPECT_ACTION(otMacFrameDoesAddrMatch(&sReceiveFrame, sPanid, sShortAddress, &sExtAddress),
                     error = OT_ERROR_ABORT);
 
+    iid = getIidFromFrame(&sReceiveFrame);
+    assert(iid != INVALID_VALUE);
```
Comment (hemanth-silabs) [inline examples/platforms/simulation/radio.c:1123]: Same as above.

**Review by hemanth-silabs — COMMENTED** [review]
LGTM 👍

**Review by suveshpratapa — APPROVED** [review]
LGTM

File: examples/platforms/simulation/radio.c (hunk 1 above)
Comment (parag-silabs) [reply examples/platforms/simulation/radio.c:341]: Agree, but since there is no test at this point for multipan for wrong dest panId I purposely added it to catch that unexpected frame. I thought we should remove this check whenever we purposely allow wrong dest panId, do we happen to know any use case that it might happen? I am happy to change it if it makes sense to remove this check though; let me know. I think simulated radio filtered the packets based on current and broadcast PANID so we should not be getting wrong dest panId unless radio in promiscuous mode.

For promiscuous mode, I think simulated radio allows all the packets but that code flow does not go through this function. 

For non multipan cases, iid will be always zero, so this should not affect the existing tests.

File: examples/platforms/simulation/radio.c (hunk 2 above)
Comment (parag-silabs) [reply examples/platforms/simulation/radio.c:1123]: Same response as above. Please let me know what you think.

File: examples/platforms/simulation/radio.c [hunk 3]
```diff
@@ -299,25 +308,49 @@ static void ReverseExtAddress(otExtAddress *aReversed, const otExtAddress *aOrig
...
+    otPanId destPanId;
+
+    destPanId = otMacGetDstPanId(aFrame);
+    if (destPanId == RADIO_BCAST_PANID)
+    {
+        iid = RADIO_BCAST_IID;
+    }
+    else
+    {
+        iid = utilsSoftSrcMatchFindIidFromPanId(destPanId);
```
[the remark this answers is not in the document]
Comment (parag-silabs) [reply examples/platforms/simulation/radio.c:325]: I've purposely not handled it so we can flag such packets (see my other comment for more info). Also to let the caller decides if it is an invalid iid.

File: examples/platforms/simulation/radio.c (hunk 1 above)
Comment (hemanth-silabs) [reply examples/platforms/simulation/radio.c:341]: Good point!!

File: script/test [hunk 4]
```diff
@@ -86,6 +86,9 @@ readonly FULL_LOGS
 TREL=${TREL:-0}
 readonly TREL
 
+MULTIPAN_RCP="${MULTIPAN_RCP:-1}"
```
Comment (bertoldi-silabs) [inline script/test:89]: Does it make sense to default this option to 1, or is this option even needed in the script at all? This will effect all of the workflow options that run the test script to build the rcp simulation/posix libraries.

Another option could be to add in a new stage, maybe in the posix.yml workflow, to run one stage of tests with OT_MULTIPAN_RCP=ON and the OT_BUILDDIR=${PWD}/build/multipan (or something like that?). 

I think that this would simplify the amount of changes that need to be made for these tests to be run in the workflow and minimize the impact on other stages. However, would this increase the complexity needed to access the apps in the expect test you wrote?

**Review by hemanth-silabs — APPROVED** [review]
LGTM 👍

File: script/test (hunk 4 above)
Comment (parag-silabs) [reply script/test]: Make sense to disable it by default. Just pushed the changes.

File: tests/scripts/expect/posix-rcp-non-zero-iid.exp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+# This test uses posix ot-cli and ot-rcp app with non-zero spinel iid support.
+# There are three nodes in this test (leader, rx-on-when-idle and rx-off-when-idle )
+# and each node spawn with non-zero iid.
+# This test perform energy scan, form and join the Thread network followed by ping
+# operation from a leader to each of the node.
+
+source "tests/scripts/expect/_common.exp"
+source "tests/scripts/expect/_multinode.exp"
+
+set is_mp_enabled $::env(MULTIPAN_RCP)
```
Comment (bertoldi-silabs) [inline tests/scripts/expect/posix-rcp-non-zero-iid.exp:39]: I think that it is useful to have an option to skip running the test if it isn't configured to run but still executed; however, this command fails when I run on my machine without MULTIPAN_RCP defined:

```
no such variable  
    (read trace on "::env(MULTIPAN_RCP)")  
    invoked from within  
"set is_mp_enabled $::env(MULTIPAN_RCP)"  
    (file "./tests/scripts/expect/posix-rcp-non-zero-iid.exp" line 39)
```

For your consideration, this change on line 39 seems to work for me:

```
-set is_mp_enabled $::env(MULTIPAN_RCP)  
+set is_mp_enabled [info exists env(MULTIPAN_RCP)]
```

Comment (bertoldi-silabs) [reply tests/scripts/expect/posix-rcp-non-zero-iid.exp:39]: On further inspection the environment variable is exported in `script/test` regardless of if MULTIPAN_RCP is enabled. The error that I see will still happen if the environment variable is not set, but there are several other variables that will cause similar errors, which implies that these tests are expected to be invoked by `script/test` or the environment variables must be manually set.

**Review by bertoldi-silabs — APPROVED** [review]

**Review by lmnotran — APPROVED** [review]
